util/auto_tune_compressor.cc

```diff
@@ -144,138 +153,211 @@ CostAwareCompressor::CostAwareCompressor(const CompressionOptions& opts)
+                                  local_wa);
+  } else {
+    auto chosen_index = SelectCompressionBasedOnGoal(local_wa);
+    // Check if the chosen compression type and level are available
+    // if not, skip the compression
+    if (chosen_index >= compressors_.size()) {
+      *out_compression_type = kNoCompression;
+      return Status::OK();
+    }
+    return CompressBlockAndRecord(chosen_index, uncompressed_data,
+                                  compressed_output, out_compression_type,
+                                  local_wa);
+  }
 }
 
-Compressor::ManagedWorkingArea CostAwareCompressor::ObtainWorkingArea() {
-  auto wrap_wa = allcompressors_.back().back()->ObtainWorkingArea();
+Compressor::ManagedWorkingArea
+AutoCompressionAlgoLevelSelector::ObtainWorkingArea() {
+  auto wrap_wa = compressors_.back()->ObtainWorkingArea();
   auto wa = new CostAwareWorkingArea(std::move(wrap_wa));
   // Create cost predictors for each compression type and level
-  wa->cost_predictors_.reserve(allcompressors_.size());
-  for (size_t i = 0; i < allcompressors_.size(); i++) {
-    CompressionType type = static_cast<CompressionType>(i + 1);
-    if (allcompressors_[type - 1].size() == 0) {
-      wa->cost_predictors_.emplace_back();
-      continue;
-    } else {
-      std::vector<IOCPUCostPredictor*> predictors_diff_levels;
-      predictors_diff_levels.reserve(kCompressionLevels[type - 1].size());
-      for (size_t j = 0; j < kCompressionLevels[type - 1].size(); j++) {
-        predictors_diff_levels.emplace_back(new IOCPUCostPredictor(10));
-      }
-      wa->cost_predictors_.emplace_back(std::move(predictors_diff_levels));
+  wa->cost_predictors_.reserve(compressors_.size());
+  for (size_t i = 0; i < kCompressionLevels.size(); i++) {
+    for (size_t j = 0; j < kCompressionLevels[i].size(); j++) {
+      wa->cost_predictors_.emplace_back(new IOCPUCostPredictor(kWindow));
     }
   }
   return ManagedWorkingArea(wa, this);
 }
-void CostAwareCompressor::ReleaseWorkingArea(WorkingArea* wa) {
+void AutoCompressionAlgoLevelSelector::MeasureUtilization() {
+  usage_tracker_.Record();
+}
+void AutoCompressionAlgoLevelSelector::ReleaseWorkingArea(WorkingArea* wa) {
   // remove all created cost predictors
-  for (auto& prdictors_diff_levels :
+  for (auto& predictor :
        static_cast<CostAwareWorkingArea*>(wa)->cost_predictors_) {
-    for (auto& predictor : prdictors_diff_levels) {
-      delete predictor;
-    }
+    delete predictor;
   }
   delete static_cast<CostAwareWorkingArea*>(wa);
 }
+size_t AutoCompressionAlgoLevelSelector::SelectCompressionBasedOnGoal(
+    CostAwareWorkingArea* wa) {
+  // If no budgets are available, use default choice
+  if (!cpu_budget_ || !io_budget_) {
+    size_t default_choice = 0;
+    return default_choice;
+  }
+  if ((block_count_.fetch_add(1, std::memory_order_relaxed)) %
+          kDecideEveryNBlocks !=
+      0) {
+    return cur_compressor_idx_;
+  }
```

Comment:
This seems reductant to kExplorationPercentage. Isn't kDecideEveryNBlocks like kExplorationPercentage for auto tune?

Reply:
I see kDecideEveryNBlocks should be a noun not a phrase. Maybe something like kCompressionEvaluationInterval